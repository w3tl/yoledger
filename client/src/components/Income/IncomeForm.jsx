import React from 'react';
import PropTypes from 'prop-types';
import { TextInput, Button } from '../Elements';
import { incomePropType } from './propTypes';

class IncomeForm extends React.Component {
  constructor(props) {
    super(props);
    const { income } = props;
    this.state = {
      isCreate: !income.id,
      name: income.name,
    };
  }

  handleChange = name => ({ target }) => {
    this.setState({
      [name]: target.value,
    });
  }

  handleSubmit = (e) => {
    e.preventDefault();
    const { name } = this.state;
    const { onSave } = this.props;
    onSave({
      name,
    });
  }

  render() {
    const { isCreate, name } = this.state;
    return (
      <form onSubmit={this.handleSubmit}>
        {isCreate && <TextInput label="Name" value={name} onChange={this.handleChange('name')} />}
        {!isCreate && <h4>{name}</h4>}
        {isCreate && (
          <Button disabled={name.length === 0} type="submit">
            Save
          </Button>)}
      </form>
    );
  }
}

IncomeForm.propTypes = {
  income: incomePropType,
  onSave: PropTypes.func,
};
IncomeForm.defaultProps = {
  income: {
    id: null,
    name: '',
  },
  onSave: () => {},
};

export default IncomeForm;
