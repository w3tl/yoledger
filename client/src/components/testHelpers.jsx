import React from 'react';
import { shallow, mount } from 'enzyme';
import { MockedProvider } from 'react-apollo/test-utils';

export const wait = (ms = 0) => new Promise(resolve => setTimeout(resolve, ms));

export const withProvider = Component => props => (
  <MockedProvider {...props}>
    <Component />
  </MockedProvider>
);

export const testRenderWithoutError = (Component) => {
  it('renders without error', async () => {
    const wrapper = shallow(Component);
    await wait();
    wrapper.update();
    expect(wrapper).toMatchSnapshot();
  });
};

export const testLoadingState = (Component) => {
  it('should render loading state initially', () => {
    const wrapper = mount(Component);
    expect(wrapper.text()).toEqual('Loading...');
  });
};

export const testErrorUI = (Component) => {
  it('should show error UI', async () => {
    const wrapper = mount(Component);
    await wait(0); // wait for response
    expect(wrapper.update().text()).toContain('Error');
  });
};

// THNX: http://blog.blakesimpson.co.uk/read/98-how-to-stub-date-in-jest-tests-building-a-stubdate-helper
export const stubDate = (fixedDate) => {
  let originalDate;

  beforeAll(() => {
    originalDate = Date;

    global.Date = class extends Date {
      constructor() {
        super();

        return fixedDate;
      }
    };
  });

  afterAll(() => {
    global.Date = originalDate;
  });
};
