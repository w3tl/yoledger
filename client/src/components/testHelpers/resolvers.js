import dateResolver from '../../../../graphql/date';
import assets from '../Asset/mockData';
import expenses from '../Expense/mockData';
import incomes from '../Income/mockData';
import { accounts, periods, budgets } from '../Budget/mockData';

export default {
  Date: dateResolver,
  Query: {
    account(root, { id }) {
      if (id === '666') return { name: 'new account' };
      return assets.concat(expenses).concat(incomes).find(a => a.id === id);
    },
    accounts(root, { type }) {
      if (type === 'ASSET') return assets;
      if (type === 'EXPENSE') return expenses;
      if (type === 'INCOME') return incomes;
      return [];
    },
    budgets() {
      return { accounts, periods: periods.map(dateStr => new Date(dateStr)) };
    },
    budget() {
      return budgets[0];
    },
  },
  Mutation: {
    upsertBudget(root, { input: { date } }) {
      return {
        success: true,
        allocation: {
          id: date.toISOString().concat('new account'),
          account: { id: '99', name: 'new account' },
        },
      };
    },
    addAccount(root, { input }) {
      return {
        account: { id: '666', name: input.name },
      };
    },
  },
};
