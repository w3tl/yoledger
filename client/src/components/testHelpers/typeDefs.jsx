export default `
scalar Date
type Query {
  account(id: ID!): Account
  accounts(type: AccountType!): [Account]

  budgets(dateStart: Date!, count: Int = 5): Budget
  budget(date: Date!): [Allocation]
}
type Mutation {
  addAccount(input: AddAccountInput!): AddAccountPayload
  upsertBudget(input: UpsertBudgetInput!): UpsertBudgetPayload
}

type Account {
  id: ID!
  name: String!
  balance: Float
  type: AccountType!
}

type Budget {
  periods: [Date]
  accounts: [Account]
}

type Allocation {
  id: String!
  account: Account!
  amount: Float
  balance: Float
}

enum AccountType {
  ASSET
  EXPENSE
  INCOME
  VENDOR
}

input AddAccountInput {
  name: String!
  balance: Float
  type: AccountType!
}

type AddAccountPayload {
  account: Account
}

input UpsertBudgetInput {
  account: String!
  date: Date!
  amount: Float!
}

type UpsertBudgetPayload {
  success: Boolean
  allocation: Allocation
}
`;
