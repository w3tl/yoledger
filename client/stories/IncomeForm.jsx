import React from 'react';

import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import IncomeForm from '../src/components/Income/IncomeForm';
import incomes from '../src/components/Income/mockData';

storiesOf('Income/Form', module)
  .add('create', () => <IncomeForm onSave={action('onSave')} />)
  .add('edit', () => <IncomeForm income={incomes[0]} onSave={action('onSave')} />);
