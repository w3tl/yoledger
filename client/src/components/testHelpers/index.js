import React from 'react';
import { shallow, mount } from 'enzyme';
import { ApolloProvider } from 'react-apollo';
import { ApolloClient } from 'apollo-client';
import { InMemoryCache } from 'apollo-cache-inmemory';
import { SchemaLink } from 'apollo-link-schema';
import { makeExecutableSchema } from 'graphql-tools';
import typeDefs from './typeDefs';
import resolvers from './resolvers';

export const wait = (ms = 0) => new Promise(resolve => setTimeout(resolve, ms));

const executableSchema = makeExecutableSchema({
  typeDefs,
  resolvers,
});

const cache = new InMemoryCache();
const link = new SchemaLink({ schema: executableSchema });
export const client = new ApolloClient({
  link,
  cache,
});

export const withProvider = Component => props => (
  <ApolloProvider client={client}>
    <Component client={client} />
  </ApolloProvider>
);

export const testRenderWithoutError = (Component) => {
  it('renders without error', async () => {
    const wrapper = shallow(Component);
    await wait();
    wrapper.update();
    expect(wrapper).toMatchSnapshot();
  });
};

export const testLoadingState = (Component) => {
  it('should render loading state initially', () => {
    const wrapper = mount(Component);
    expect(wrapper.text()).toEqual('Loading...');
  });
};

export const testErrorUI = (Component) => {
  it('should show error UI', async () => {
    const wrapper = mount(Component);
    await wait(0); // wait for response
    expect(wrapper.update().text()).toContain('Error');
  });
};

// THNX: http://blog.blakesimpson.co.uk/read/98-how-to-stub-date-in-jest-tests-building-a-stubdate-helper
export const stubDate = (fixedDate) => {
  const date = new Date(fixedDate);
  let originalDate;

  beforeAll(() => {
    originalDate = Date;

    global.Date = class extends Date {
      constructor() {
        super();

        return date;
      }
    };
  });

  afterAll(() => {
    global.Date = originalDate;
  });
};
