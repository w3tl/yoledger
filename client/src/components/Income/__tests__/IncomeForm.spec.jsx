import React from 'react';
import { mount } from 'enzyme';
import IncomeForm from '../IncomeForm';
import { testRenderWithoutError } from '../../testHelpers';
import incomes from '../mockData';

const income = incomes[0];

describe('IncomeForm', () => {
  // TODO: test create mode
  testRenderWithoutError(<IncomeForm income={income} />);

  it('should handle submit on edit mode', () => {
    const onSave = jest.fn();
    const output = mount(
      <IncomeForm income={income} onSave={onSave} />,
    );
    expect(output.state('isCreate')).toBeFalsy();
    output.find('form').simulate('submit');
    expect(onSave).toHaveBeenCalledWith({
      name: income.name,
    });
  });

  it('should handle submit on create mode', () => {
    const onSave = jest.fn();
    const output = mount(
      <IncomeForm onSave={onSave} />,
    );
    expect(output.state('isCreate')).toBeTruthy();
    output.find('input').simulate('change', { target: { value: 'new account' } });
    output.find('form').simulate('submit');
    expect(onSave).toHaveBeenCalledWith({
      name: 'new account',
    });
  });
});
