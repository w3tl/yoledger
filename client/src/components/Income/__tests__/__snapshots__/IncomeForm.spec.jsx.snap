// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IncomeForm renders without error 1`] = `
<form
  onSubmit={[Function]}
>
  <h4>
    Salary
  </h4>
  <Button
    type="submit"
  >
    Save
  </Button>
</form>
`;
